fe/fe-core/src/main/java/com/starrocks/meta/lock/LockManager.java

```diff
@@ -0,0 +1,521 @@
+            lockTables[i] = new HashMap<>();
+        }
+    }
+
+    /**
+     * Attempt to acquire a lock of 'lockType' on rid
+     *
+     * @param rid      The resource id to lock
+     * @param locker   The Locker to lock this on behalf of.
+     * @param lockType Then lock type requested
+     * @param timeout  milliseconds to time out after if lock couldn't be obtained. 0 means block indefinitely.
+     *
+     * @throws LockTimeoutException when the transaction time limit was exceeded.
+     * @throws DeadlockException when deadlock was detected
+     */
+
+    public void lock(long rid, Locker locker, LockType lockType, long timeout)
+            throws LockInterruptException, LockTimeoutException, DeadlockException {
+
+        final long startTime;
+
+        synchronized (locker) {
+            int lockTableIdx = getLockTableIndex(rid);
+            synchronized (lockTableMutexes[lockTableIdx]) {
+                Map<Long, Lock> lockTable = lockTables[lockTableIdx];
+                Lock lock = lockTable.get(rid);
+
+                if (lock == null) {
+                    lock = new LightWeightLock();
+                    lockTable.put(rid, lock);
+                } else if (lock instanceof LightWeightLock) {
+                    List<LockHolder> owners = new ArrayList<>(lock.getOwners());
+                    if (owners.size() > 0) {
+                        /* Lock is already held by someone else so mutate. */
+                        lock = new MultiUserLock(owners.get(0));
+                        lockTable.put(rid, lock);
+                    }
+                }
+
+                LockGrantType lockGrantType = lock.lock(locker, lockType);
+                if (lockGrantType == LockGrantType.NEW || lockGrantType == LockGrantType.EXISTING) {
+                    return;
+                }
+            }
+
+            startTime = System.currentTimeMillis();
+            locker.setWaitingFor(rid, lockType);
+
+            /*
+             * Because deadlock detection also requires a significant cost, but at the first moment
+             * when a lock cannot be obtained, we cannot determine whether it is because the required
+             * lock is being used normally or if a deadlock has occurred.
+             * Therefore, based on the configuration parameter `dead_lock_detection_delay_time_ms`
+             * is used to control the waiting time before deadlock detection.
+             * If a lock is obtained during this period, there is no need to perform deadlock detection.
+             * Avoid frequent and unnecessary deadlock detection due to lock contention
+             */
+            long deadLockDetectionDelayTimeMs = Config.dead_lock_detection_delay_time_ms;
+            if (Config.dead_lock_detection && deadLockDetectionDelayTimeMs > 0) {
+                if (timeout != 0) {
+                    deadLockDetectionDelayTimeMs = Math.min(deadLockDetectionDelayTimeMs, timeRemain(timeout, startTime));
```

Comment:
If we have tryLock(timeout = 100ms), the dead lock detection will be called too frequently.